### PR TITLE
Next Nonce Issue

### PR DESCRIPTION
**Pull request by otherview**
There's an issue with 
```
func NextNonce(cl *obscuroclient.Client, w wallet.Wallet) uint64 {
	// only returns the nonce when the previous transaction was recorded
	for {
		result := readNonce(cl, w.Address())
		if result == w.GetNonce() {
			return w.GetNonceAndIncrement()
		}
		time.Sleep(time.Millisecond)
	}
}
```

If this breaks for some reason, no other transaction goes through (for that wallet).
If no other tx goes through that's fine because the validate_chain doesn't expect any tx.
But this creates a go routine that never ends.

### CODE REVIEW

File: integration/simulation/transaction_injector.go [hunk 1]
```diff
@@ -346,13 +346,14 @@ func readNonce(cl *obscuroclient.Client, a common.Address) uint64 {
 	return result
 }
 
+// NextNonce waits for the previous transaction to be recorded - returns a value if a timeout has passed
 func NextNonce(cl *obscuroclient.Client, w wallet.Wallet) uint64 {
-	// only returns the nonce when the previous transaction was recorded
-	for {
+	for startTime := time.Now(); time.Since(startTime) < 5*time.Second; time.Sleep(time.Millisecond) {
```
Comment (tudor-malene) [inline integration/simulation/transaction_injector.go:351]: probably not best to measure this in seconds but in avgBlock Time units

**Review by tudor-malene — CHANGES_REQUESTED** [review]
one change needed, since this will break tests that run with a larger avg block time

File: integration/simulation/transaction_injector.go (hunk 1 above)
Comment (otherview) [reply integration/simulation/transaction_injector.go]: Good point 👍

File: integration/simulation/transaction_injector.go [hunk 2]
```diff
@@ -311,21 +318,6 @@ func (m *TransactionInjector) rndL2NodeClient() *obscuroclient.Client {
 	return m.l2Clients[rand.Intn(len(m.l2Clients))] //nolint:gosec
 }
 
-func newObscuroTransferTx(from wallet.Wallet, dest common.Address, amount uint64, client *obscuroclient.Client) types.TxData {
```
Comment (tudor-malene) [inline integration/simulation/transaction_injector.go:314]: why did you remove these functions?
I think they were adding clarity to what's going on

**Review by tudor-malene — CHANGES_REQUESTED** [review]

File: integration/simulation/transaction_injector.go (hunk 2 above)
Comment (otherview) [reply integration/simulation/transaction_injector.go:314]: I don't really agree.. They seem to be doing one instruction, the tx data generation.

And it's starting to have super similar instructions like `newCustomObscuroWithdrawalTx` and `newObscuroWithdrawalTx` which is only one more step to follow, to figure out what it's doing.

Having to make then a part of the struct feels super bloated.

**Comment by otherview** [comment]
Moved this to an issue
